### PR TITLE
Proxy: Use trust-dns-resolver for DNS.

### PR DESCRIPTION
**Pull request by briansmith**
trust-dns-resolver is a more complete implementation. In particular,
it supports CNAMES correctly, which is needed for PR #764. It also
supports /etc/hosts, which will help with issue #62.

Use the 0.8.2 pre-release since it hasn't been released yet. It was
created at our request.

Signed-off-by: Brian Smith <brian@briansmith.org>

### CODE REVIEW

File: proxy/Cargo.toml [hunk 1]
```diff
@@ -36,8 +35,7 @@ tokio-signal = "0.1"
 prost = "0.3.0"
 prost-types = "0.3.0"
 
-abstract-ns = "0.4"
-ns-dns-tokio = "0.4"
+trust-dns-resolver = { version = "0.8.2", default-features = false, git = "https://github.com/bluejekyll/trust-dns", branch = "0.8_release" }
```
Comment (briansmith) [inline proxy/Cargo.toml]: @bluejekyll will release trust-dns-resolver 0.8.2 off this branch soon. This branch was created, in part, so we wouldn't need to deal with any hassles associated with the new tokio, which trust-dns-resolver 0.9 uses.

File: proxy/src/dns.rs [hunk 2]
```diff
@@ -1,68 +1,61 @@
...
-        //  TODO: Avoid this extra allocation.
-        let s = s.to_ascii_lowercase();
-        abstract_ns::Name::from_str(&s)
-            .map(Name)
-            .map_err(|_| ())
+impl<'a> From<&'a str> for Name {
+    fn from(s: &str) -> Self {
+        // TODO: Verify the name is a valid DNS name.
+        // TODO: Avoid this extra allocation.
+        Name(s.to_ascii_lowercase())
```
Comment (briansmith) [inline proxy/src/dns.rs]: My understanding is the http create will validate these names for us before we extract them. Our use of `abstract_ns::Name::from_str()` for validation was misguided (based on discussion with the maintainer of abstract_ns) and didn't catch all potential issues anyway.

File: proxy/src/dns.rs [hunk 3]
```diff
@@ -73,70 +66,92 @@ impl AsRef<str> for Name {
...
 
 impl Config {
-    /// Note that this ignores any errors reading or parsing the resolve.conf
-    /// file, just like the `domain` crate does.
-    pub fn from_file(resolve_conf_path: &Path) -> Self {
-        let mut resolv_conf = domain::resolv::ResolvConf::new();
-        let _ = resolv_conf.parse_file(resolve_conf_path);
-        resolv_conf.finalize();
-        Config(resolv_conf)
+    /// TODO: Make this infallible, like it is in the `domain` crate.
```
Comment (briansmith) [inline proxy/src/dns.rs]: This is one of the bigger risks in this PR. The `domain` crate used a parser for /etc/resolv.conf that was infallible; trust-dns-resolver doesn't have an infallible parser. OTOH, parsing junk input is pretty dangerous and difficult to reason about. In practice we expect Kubernetes will write valid configuration files.

Comment (hawkw) [reply proxy/src/dns.rs]: To me, it seems that this function being falliable is more correct --- what would happen if the `domain` parser were to read junk input? I suspect that whatever the answer to that question is, it's probably worse than unwrapping this result & bailing with a descriptive message.

File: proxy/Cargo.toml (hunk 1 above)
Comment (bluejekyll) [reply proxy/Cargo.toml]: fyi: https://www.crates.io/crates/trust-dns-resolver

**Comment by olix0r** [comment]
@briansmith Can you please merge master when you have a chance? Your branch has a telemetry bug that was fixed on master that makes it hard to test this completely. (I'm merging it myself and verifying now, but it looks like you have a conflict anyway)

**Comment by olix0r** [comment]
I'm not sure what's changed between your branch and master, but after merging i see:

```
   Compiling conduit-proxy v0.3.0 (file:///Users/ver/s/go/src/github.com/runconduit/conduit/proxy)
error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> proxy/src/control/observe.rs:50:32
   |
50 |             return future::err(grpc::Error::Grpc(grpc::Status::INTERNAL));
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 parameters

error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> proxy/src/control/observe.rs:59:36
   |
59 |                   return future::err(grpc::Error::Grpc(
   |  ____________________________________^
60 | |                     grpc::Status::INVALID_ARGUMENT,
61 | |                 ));
   | |_________________^ expected 2 parameters

error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> proxy/src/control/observe.rs:73:36
   |
73 |                 return future::err(grpc::Error::Grpc(grpc::Status::INTERNAL));
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 parameters

error: aborting due to 3 previous errors

error: Could not compile `conduit-proxy`.
```

But I don't see any changes to the `tower_grpc` dependency...

File: proxy/src/lib.rs [hunk 4]
```diff
@@ -203,7 +201,11 @@ where
             config.event_buffer_capacity,
         );
 
-        let dns_config = dns::Config::from_file(&config.resolv_conf_path);
+        let dns_config = dns::Config::from_system_config()
+            .unwrap_or_else(|e| {
+                // TODO: Make DNS configuration infallible.
+                panic!("invalid DNS configuration: {:?}", e);
+            });
```
[the remark this answers is not in the document]
Comment (briansmith) [reply proxy/src/lib.rs]: > tioli: isn't this equivalent to .expect("invalid DNS configuration")?

It is similar, except in the variant I wrote, we print the error encountered.

**Pull request merged**